toolchain/parse/handle_decl_name_and_params.cpp

```diff
@@ -6,118 +6,85 @@
 
 namespace Carbon::Parse {
 
-// Handles DeclNameAndParamsAs(Optional|Required).
-static auto HandleDeclNameAndParams(Context& context, State after_name)
-    -> void {
+auto HandleDeclNameAndParams(Context& context) -> void {
   auto state = context.PopState();
 
-  // TODO: Should handle designated names.
-  if (auto identifier = context.ConsumeIf(Lex::TokenKind::Identifier)) {
-    context.PushState(state, after_name);
-
-    if (context.PositionIs(Lex::TokenKind::Period)) {
-      context.AddLeafNode(NodeKind::IdentifierName, *identifier);
-      context.PushState(state, State::PeriodAsDecl);
-    } else {
-      context.AddLeafNode(NodeKind::IdentifierName, *identifier);
-    }
-  } else {
+  auto identifier = context.ConsumeIf(Lex::TokenKind::Identifier);
+  if (!identifier) {
     CARBON_DIAGNOSTIC(ExpectedDeclName, Error,
                       "`{0}` introducer should be followed by a name.",
                       Lex::TokenKind);
+    CARBON_DIAGNOSTIC(ExpectedDeclNameAfterPeriod, Error,
+                      "`.` should be followed by a name.");
     Lex::TokenIndex token = *context.position();
     if (context.tokens().GetKind(token) == Lex::TokenKind::FileEnd) {
       // The end of file is an unhelpful diagnostic location. Instead, use the
       // introducer token.
       token = state.token;
     }
-    context.emitter().Emit(token, ExpectedDeclName,
-                           context.tokens().GetKind(state.token));
+    if (state.token == *context.position()) {
+      context.emitter().Emit(token, ExpectedDeclNameAfterPeriod);
+    } else {
+      context.emitter().Emit(token, ExpectedDeclName,
+                             context.tokens().GetKind(state.token));
+    }
```

Comment:
As you're restructuring this, can you keep the diagnostic definitions with where they're used?

Reply:
Done.